File: injector/manifest_patcher.py
"""
AdSweep Injector - Binary AndroidManifest.xml patcher.

Modifies the binary AXML manifest inside a built APK without
decompiling resources. Uses direct byte manipulation on the
Android binary XML format.
"""
import os
import struct
import zipfile


# Android resource IDs for attributes we want to modify
RES_EXTRACT_NATIVE_LIBS = 0x010104ea  # android:extractNativeLibs
RES_IS_SPLIT_REQUIRED = 0x01010591    # android:isSplitRequired


def patch_manifest_in_apk(apk_path: str) -> bool:
    """Patch the binary AndroidManifest.xml inside an APK.

    Supports:
    - Setting extractNativeLibs to true
    - Setting isSplitRequired to false
    - Removing requiredSplitTypes attribute
    - Removing com.android.vending.splits.required meta-data
    """
    try:
        with zipfile.ZipFile(apk_path) as z:
            data = bytearray(z.read("AndroidManifest.xml"))
    except Exception as e:
        print(f"[!] Cannot read manifest from APK: {e}")
        return False

    modified = False

    # Patch extractNativeLibs = true
    if _set_boolean_attribute(data, RES_EXTRACT_NATIVE_LIBS, True):
        print("[+] Set extractNativeLibs=true in binary manifest")
        modified = True

    # Patch isSplitRequired = false
    if _set_boolean_attribute(data, RES_IS_SPLIT_REQUIRED, False):
        print("[+] Set isSplitRequired=false in binary manifest")
        modified = True

    # Remove requiredSplitTypes string attribute
    patched = _remove_string_attribute(data, "requiredSplitTypes")
    if patched is not None:
        data = patched
        print("[+] Removed requiredSplitTypes from binary manifest")
        modified = True

    # Remove splitTypes string attribute
    patched = _remove_string_attribute(data, "splitTypes")
    if patched is not None:
        data = patched
        print("[+] Removed splitTypes from binary manifest")
        modified = True

    if not modified:
        return True  # Nothing to change

    # Write back to APK
    try:
        _replace_in_zip(apk_path, "AndroidManifest.xml", bytes(data))
        return True
    except Exception as e:
        print(f"[!] Failed to write manifest back to APK: {e}")
        return False


def _set_boolean_attribute(data: bytearray, res_id: int, value: bool) -> bool:
    """Find a boolean attribute by resource ID and change its value.

    In binary AXML format:
    - Resource IDs are mapped to string pool indices via the resource map chunk
    - Attribute entries are 20 bytes: ns(4) + name(4) + rawValue(4) + typedSize(2) + res0(1) + type(1) + data(4)
    - Boolean type = 0x12, false = 0x00000000, true = 0xFFFFFFFF
    """
    # Find the resource map to get the string pool index for this resource ID
    string_index = _find_string_index_for_res_id(data, res_id)
    if string_index < 0:
        return False

    # Find attribute entries referencing this string index with boolean type
    name_bytes = struct.pack('<I', string_index)
    target_data = 0xFFFFFFFF if value else 0x00000000
    found = False
    offset = 0

    while True:
        idx = data.find(name_bytes, offset)
        if idx < 0:
            break

        if idx + 16 <= len(data):
            dtype = data[idx + 11]
            if dtype == 0x12:  # TYPE_INT_BOOLEAN
                current = struct.unpack_from('<I', data, idx + 12)[0]
                if current != target_data:
                    struct.pack_into('<I', data, idx + 12, target_data)
                    found = True

        offset = idx + 4

    return found


def _find_string_index_for_res_id(data: bytearray, res_id: int) -> int:
    """Find the string pool index that maps to a given Android resource ID."""
    # Resource map chunk type = 0x0180
    res_map_type = struct.pack('<H', 0x0180)
    rm_idx = data.find(res_map_type)
    if rm_idx < 0:
        return -1

    rm_header_size = struct.unpack_from('<H', data, rm_idx + 2)[0]
    rm_size = struct.unpack_from('<I', data, rm_idx + 4)[0]
    rm_count = (rm_size - rm_header_size) // 4

    for i in range(rm_count):
        rid = struct.unpack_from('<I', data, rm_idx + rm_header_size + i * 4)[0]
        if rid == res_id:
            return i

    return -1


def _remove_string_attribute(data: bytearray, attr_name: str) -> bytearray:
    """Remove a string attribute by zeroing out its name in the string pool.

    Parses the AXML string pool, finds the attribute name string,
    and overwrites it with underscores to effectively disable it.
    """
    if len(data) < 28:
        return None

    # Parse header
    chunk_type = struct.unpack_from('<H', data, 0)[0]

    # Find string pool offset
    sp_offset = 8  # default for standalone string pool
    if chunk_type == 0x0003:  # ResXMLTree
        header_size = struct.unpack_from('<H', data, 2)[0]
        sp_offset = header_size
        sp_type = struct.unpack_from('<H', data, sp_offset)[0]
        if sp_type != 0x0001:
            return None

    sp_header_size = struct.unpack_from('<H', data, sp_offset + 2)[0]
    string_count = struct.unpack_from('<I', data, sp_offset + 8)[0]
    flags = struct.unpack_from('<I', data, sp_offset + 16)[0]
    strings_start = struct.unpack_from('<I', data, sp_offset + 20)[0]

    is_utf8 = (flags & (1 << 8)) != 0
    offsets_start = sp_offset + 28
    str_data_start = sp_offset + strings_start

    for i in range(string_count):
        str_offset = struct.unpack_from('<I', data, offsets_start + i * 4)[0]
        abs_offset = str_data_start + str_offset

        s = _read_string_at(data, abs_offset, is_utf8)
        if s == attr_name:
            _write_underscores_at(data, abs_offset, is_utf8)
            return data

    return None


def _read_string_at(data: bytearray, offset: int, is_utf8: bool) -> str:
    """Read a string from the AXML string pool at the given offset."""
    try:
        if is_utf8:
            # char length (1 or 2 bytes)
            char_len = data[offset] & 0xFF
            offset += 1
            if char_len & 0x80:
                offset += 1
            # byte length (1 or 2 bytes)
            byte_len = data[offset] & 0xFF
            offset += 1
            if byte_len & 0x80:
                byte_len = ((byte_len & 0x7F) << 8) | (data[offset] & 0xFF)
                offset += 1
            return data[offset:offset + byte_len].decode('utf-8')
        else:
            char_len = struct.unpack_from('<H', data, offset)[0]
            offset += 2
            chars = struct.unpack_from(f'<{char_len}H', data, offset)
            return ''.join(chr(c) for c in chars)
    except Exception:
        return ""


def _write_underscores_at(data: bytearray, offset: int, is_utf8: bool):
    """Overwrite a string in the pool with underscores (same length)."""
    try:
        if is_utf8:
            char_len = data[offset] & 0xFF
            skip = 1
            if char_len & 0x80:
                skip = 2
            offset += skip
            byte_len = data[offset] & 0xFF
            skip = 1
            if byte_len & 0x80:
                byte_len = ((byte_len & 0x7F) << 8) | (data[offset + 1] & 0xFF)
                skip = 2
            offset += skip
            for i in range(byte_len):
                data[offset + i] = ord('_')
        else:
            char_len = struct.unpack_from('<H', data, offset)[0]
            offset += 2
            for i in range(char_len):
                struct.pack_into('<H', data, offset + i * 2, ord('_'))
    except Exception:
        pass


def _replace_in_zip(zip_path: str, entry_name: str, new_data: bytes):
    """Replace a single entry in a ZIP file."""
    tmp_path = zip_path + ".tmp"
    with zipfile.ZipFile(zip_path, 'r') as src, \
         zipfile.ZipFile(tmp_path, 'w') as dst:
        for item in src.infolist():
            if item.filename == entry_name:
                dst.writestr(item, new_data)
            else:
                dst.writestr(item, src.read(item.filename))
    os.replace(tmp_path, zip_path)

File: manager/src/main/java/com/adsweep/manager/ManifestPatcher.java
package com.adsweep.manager;

import android.util.Log;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.zip.*;

/**
 * Patches binary AndroidManifest.xml inside an APK.
 * Removes requiredSplitTypes and splitTypes attributes
 * so the APK can be installed without matching splits.
 */
public class ManifestPatcher {

    private static final String TAG = "AdSweep.Manifest";

    /**
     * Patch the manifest in an APK to remove split requirements.
     * Also sets extractNativeLibs=true.
     */
    public static boolean patch(File apkFile) {
        try {
            byte[] manifestBytes = readEntryFromZip(apkFile, "AndroidManifest.xml");
            if (manifestBytes == null) return false;

            boolean modified = false;

            // Set extractNativeLibs=true
            if (setBooleanAttribute(manifestBytes, 0x010104ea, true)) {
                modified = true;
                Log.i(TAG, "Set extractNativeLibs=true");
            }

            // Set isSplitRequired=false (allows installing without split APKs)
            if (setBooleanAttribute(manifestBytes, 0x01010591, false)) {
                modified = true;
                Log.i(TAG, "Set isSplitRequired=false");
            }

            // Remove requiredSplitTypes string attribute
            byte[] patched = removeStringAttribute(manifestBytes, "requiredSplitTypes");
            if (patched != null) {
                manifestBytes = patched;
                modified = true;
                Log.i(TAG, "Removed requiredSplitTypes");
            }

            // Remove splitTypes string attribute
            patched = removeStringAttribute(manifestBytes, "splitTypes");
            if (patched != null) {
                manifestBytes = patched;
                modified = true;
                Log.i(TAG, "Removed splitTypes");
            }

            if (modified) {
                replaceEntryInZip(apkFile, "AndroidManifest.xml", manifestBytes);
                return true;
            }
            return false;

        } catch (Exception e) {
            Log.e(TAG, "Manifest patch failed", e);
            return false;
        }
    }

    /**
     * Remove a string attribute by zeroing out its value in the string pool.
     * This effectively makes the attribute empty/ignored.
     */
    private static byte[] removeStringAttribute(byte[] data, String attrName) {
        // Find the attribute name in the string pool
        // String pool starts at offset 8 (after header type + size)
        ByteBuffer buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);

        // Parse string pool header
        int type = buf.getShort(0) & 0xFFFF; // should be 0x0001
        if (type != 0x0001 && type != 0x0003) return null;

        int headerSize = buf.getShort(2) & 0xFFFF;
        int chunkSize = buf.getInt(4);

        // For ResXMLTree (type 0x0003), string pool is the first chunk inside
        int spOffset = 8; // default
        if (type == 0x0003) {
            // Skip the XML tree header, find the string pool chunk
            spOffset = headerSize;
            int spType = buf.getShort(spOffset) & 0xFFFF;
            if (spType != 0x0001) return null; // not a string pool
        }

        int spHeaderSize = buf.getShort(spOffset + 2) & 0xFFFF;
        int spChunkSize = buf.getInt(spOffset + 4);
        int stringCount = buf.getInt(spOffset + 8);
        int styleCount = buf.getInt(spOffset + 12);
        int flags = buf.getInt(spOffset + 16);
        int stringsStart = buf.getInt(spOffset + 20);
        int stylesStart = buf.getInt(spOffset + 24);

        boolean isUtf8 = (flags & (1 << 8)) != 0;

        // Read string offsets
        int offsetsStart = spOffset + 28;
        int strDataStart = spOffset + stringsStart;

        // Search for the attribute name in string pool
        for (int i = 0; i < stringCount; i++) {
            int strOffset = buf.getInt(offsetsStart + i * 4);
            int absOffset = strDataStart + strOffset;

            String str = readStringAt(data, absOffset, isUtf8);
            if (attrName.equals(str)) {
                // Found it - zero out the string content to make it empty
                // This effectively removes the attribute by making its name unrecognizable
                // Better approach: change the string to something harmless
                writeStringAt(data, absOffset, isUtf8);
                return data;
            }
        }
        return null;
    }

    private static String readStringAt(byte[] data, int offset, boolean isUtf8) {
        try {
            if (isUtf8) {
                int charLen = data[offset] & 0xFF;
                if ((charLen & 0x80) != 0) {
                    charLen = ((charLen & 0x7F) << 8) | (data[offset + 1] & 0xFF);
                    offset++;
                }
                offset++;
                int byteLen = data[offset] & 0xFF;
                if ((byteLen & 0x80) != 0) {
                    byteLen = ((byteLen & 0x7F) << 8) | (data[offset + 1] & 0xFF);
                    offset++;
                }
                offset++;
                return new String(data, offset, byteLen, "UTF-8");
            } else {
                ByteBuffer buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
                int charLen = buf.getShort(offset) & 0xFFFF;
                offset += 2;
                char[] chars = new char[charLen];
                for (int i = 0; i < charLen; i++) {
                    chars[i] = (char) (buf.getShort(offset + i * 2) & 0xFFFF);
                }
                return new String(chars);
            }
        } catch (Exception e) {
            return "";
        }
    }

    private static void writeStringAt(byte[] data, int offset, boolean isUtf8) {
        // Overwrite string with spaces (same length, won't break structure)
        try {
            if (isUtf8) {
                int charLen = data[offset] & 0xFF;
                int skip = 1;
                if ((charLen & 0x80) != 0) { skip = 2; }
                offset += skip;
                int byteLen = data[offset] & 0xFF;
                skip = 1;
                if ((byteLen & 0x80) != 0) { skip = 2; }
                offset += skip;
                // Overwrite with underscores (safe replacement)
                for (int i = 0; i < byteLen; i++) {
                    data[offset + i] = '_';
                }
            } else {
                ByteBuffer buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
                int charLen = buf.getShort(offset) & 0xFFFF;
                offset += 2;
                for (int i = 0; i < charLen; i++) {
                    buf.putShort(offset + i * 2, (short) '_');
                }
            }
        } catch (Exception e) {}
    }

    /**
     * Set a boolean attribute by resource ID (same as Python manifest_patcher).
     */
    private static boolean setBooleanAttribute(byte[] data, int resId, boolean value) {
        ByteBuffer buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);

        // Find resource map by parsing chunk structure
        // XML tree: header(8) → StringPool(var) → ResourceMap(var) → elements...
        int rmOffset = -1;
        int offset = 8; // skip XML tree header
        while (offset < data.length - 8) {
            int t = buf.getShort(offset) & 0xFFFF;
            int size = buf.getInt(offset + 4);
            if (size <= 0 || size > data.length) break;
            if (t == 0x0180) { rmOffset = offset; break; }
            offset += size;
        }
        if (rmOffset < 0) {
            Log.w(TAG, "Resource map not found in manifest");
            return false;
        }

        int rmHeaderSize = buf.getShort(rmOffset + 2) & 0xFFFF;
        int rmSize = buf.getInt(rmOffset + 4);
        int rmCount = (rmSize - rmHeaderSize) / 4;

        // Find string index for this resource ID
        int stringIndex = -1;
        for (int i = 0; i < rmCount; i++) {
            int rid = buf.getInt(rmOffset + rmHeaderSize + i * 4);
            if (rid == resId) { stringIndex = i; break; }
        }
        if (stringIndex < 0) return false;

        // Find attribute entry with this string index and boolean type
        byte[] nameBytes = new byte[4];
        ByteBuffer.wrap(nameBytes).order(ByteOrder.LITTLE_ENDIAN).putInt(stringIndex);
        int targetData = value ? 0xFFFFFFFF : 0x00000000;
        boolean found = false;
        int searchOffset = 0;

        while (true) {
            int idx = indexOf(data, nameBytes, searchOffset);
            if (idx < 0) break;
            if (idx + 16 <= data.length) {
                int dtype = data[idx + 11] & 0xFF;
                if (dtype == 0x12) { // TYPE_INT_BOOLEAN
                    buf.putInt(idx + 12, targetData);
                    found = true;
                }
            }
            searchOffset = idx + 4;
        }
        return found;
    }

    private static int indexOf(byte[] data, byte[] pattern, int from) {
        outer:
        for (int i = from; i <= data.length - pattern.length; i++) {
            for (int j = 0; j < pattern.length; j++) {
                if (data[i + j] != pattern[j]) continue outer;
            }
            return i;
        }
        return -1;
    }

    // --- ZIP utilities ---

    private static byte[] readEntryFromZip(File zipFile, String entryName) throws Exception {
        try (ZipFile zf = new ZipFile(zipFile)) {
            ZipEntry entry = zf.getEntry(entryName);
            if (entry == null) return null;
            InputStream is = zf.getInputStream(entry);
            ByteArrayOutputStream bos = new ByteArrayOutputStream();
            byte[] buf = new byte[8192];
            int len;
            while ((len = is.read(buf)) > 0) bos.write(buf, 0, len);
            is.close();
            return bos.toByteArray();
        }
    }

    private static void replaceEntryInZip(File zipFile, String entryName, byte[] newData) throws Exception {
        File tmp = new File(zipFile.getParent(), zipFile.getName() + ".tmp");
        try (ZipFile src = new ZipFile(zipFile);
             org.apache.commons.compress.archivers.zip.ZipArchiveOutputStream zos =
                     new org.apache.commons.compress.archivers.zip.ZipArchiveOutputStream(tmp)) {
            java.util.Enumeration<? extends ZipEntry> entries = src.entries();
            while (entries.hasMoreElements()) {
                ZipEntry entry = entries.nextElement();
                byte[] data;
                if (entry.getName().equals(entryName)) {
                    data = newData;
                } else {
                    InputStream is = src.getInputStream(entry);
                    ByteArrayOutputStream bos = new ByteArrayOutputStream();
                    byte[] buf = new byte[8192];
                    int len;
                    while ((len = is.read(buf)) > 0) bos.write(buf, 0, len);
                    is.close();
                    data = bos.toByteArray();
                }
                // Preserve original compression method
                org.apache.commons.compress.archivers.zip.ZipArchiveEntry newEntry =
                        new org.apache.commons.compress.archivers.zip.ZipArchiveEntry(entry.getName());
                if (entry.getMethod() == ZipEntry.STORED) {
                    newEntry.setMethod(org.apache.commons.compress.archivers.zip.ZipArchiveEntry.STORED);
                    newEntry.setSize(data.length);
                    newEntry.setCompressedSize(data.length);
                    java.util.zip.CRC32 crc = new java.util.zip.CRC32();
                    crc.update(data);
                    newEntry.setCrc(crc.getValue());
                } else {
                    newEntry.setMethod(org.apache.commons.compress.archivers.zip.ZipArchiveEntry.DEFLATED);
                }
                zos.putArchiveEntry(newEntry);
                zos.write(data);
                zos.closeArchiveEntry();
            }
        }
        zipFile.delete();
        tmp.renameTo(zipFile);
    }
}
